### PR TITLE
Avoid pulling unnecessary blobs

### PR DESCRIPTION
**Pull request by yajo**
This will save a lot of disk and network space, and time when fetching the 1st time, mostly if your vendored repo is very big.

### CODE REVIEW

**Comment by mattmccutchen** [comment]
Thanks for the suggestion.  The concerns I can think of so far before I would want to accept this:

1. Add a test that checks in some manner that this change actually has the claimed performance benefit, both to demonstrate now that the change is useful and to catch any future regressions.  One reasonable approach might be to check that a specific blob that would be present in the cache repository with a full clone is not present with the partial clone.
2. Fully understand the effects on Braid's workflow.  At what point during a `braid update` (for example) does the missing data get downloaded?  Is it downloaded all at once, or is there a possibility of "faulting in" one promised object at a time (which could potentially be a worse performance problem than the original one this PR is addressing)?
3. Understand any potential interactions with other enhancements we're considering that might be higher priority than this one.  The obvious one to me is [this proposal](https://github.com/cristibalan/braid/issues/71#issuecomment-697295152) to avoid copying all the vendor repository data into the downstream repository.  Another such enhancement might be to use a shallow clone in the cache repository.
4. Think about version compatibility issues, both with different versions of Git that Braid claims to support (including the possibility of running multiple versions of Git on the same system) and with running multiple versions of Braid on the same system, even though we may ultimately decide we don't care about any such issues.  If it helps, raising Braid's minimum supported Git version (`REQUIRED_GIT_VERSION`) is probably an option (and probably overdue at this point), though of course, we should be aware that versions of Braid before that change is made won't enforce the new minimum.

Some of the answers may be simple and may be mainly a matter of me needing to educate myself about partial clones, which would be a good thing anyway.  I'm open to doing some or all of this work, though I'm not sure when I might get to it.  If you're willing and able to do any of the work, that may get the change accepted sooner, though I can't make promises at this point.  Of course, the primary maintainer (realityforge) may overrule me on any of this.

**Comment by Yajo** [comment]
> * Add a test that checks in some manner that this change actually has the claimed performance benefit, both to demonstrate now that the change is useful and to catch any future regressions.  One reasonable approach might be to check that a specific blob that would be present in the cache repository with a full clone is not present with the partial clone.

I don't think that's needed. You'd be testing the inner works of git, but braid is not git. This change should be so transparent that the proof that it is good would be that no tests break.

> * Fully understand the effects on Braid's workflow.  At what point during a `braid update` (for example) does the missing data get downloaded?  Is it downloaded all at once, or is there a possibility of "faulting in" one promised object at a time (which could potentially be a worse performance problem than the original one this PR is addressing)?

Check "Blobless clones" in https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/

Basically it downloads all git history, but only files in the revision they are checked out. If you move in history, it lazy-downloads missing local blobs automatically.

> * Understand any potential interactions with other enhancements we're considering that might be higher priority than this one.  The obvious one to me is [this proposal](https://github.com/cristibalan/braid/issues/71#issuecomment-697295152) to avoid copying all the vendor repository data into the downstream repository.  Another such enhancement might be to use a shallow clone in the cache repository.

Shallow clones have only partial history, so some merges can fail.

**Comment by mattmccutchen** [comment]
Belated follow-up:

> > (1) Add a test that checks in some manner that this change actually has the claimed performance benefit, both to demonstrate now that the change is useful and to catch any future regressions.  One reasonable approach might be to check that a specific blob that would be present in the cache repository with a full clone is not present with the partial clone.
> 
> I don't think that's needed. You'd be testing the inner works of git, but braid is not git. This change should be so transparent that the proof that it is good would be that no tests break.

I disagree.  For any feature we add to Braid, we want to know if it unexpectedly stops working as a result of another change, and the most practical way to do that is to add a test to the test suite.  Performance improvements are no exception: we want to know if they unexpectedly stop providing the expected benefit.  For a feature like this one, the only way to write a reliable test may be to rely on Git internals, unfortunate as it may seem.  (For example, if we tried to directly test the performance impact, we may not be able to come up with a pass/fail condition that is reliable across different machines.)

> > (2) Fully understand the effects on Braid's workflow.  At what point during a `braid update` (for example) does the missing data get downloaded?  Is it downloaded all at once, or is there a possibility of "faulting in" one promised object at a time (which could potentially be a worse performance problem than the original one this PR is addressing)?
> 
> Check "Blobless clones" in https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/
> 
> Basically it downloads all git history, but only files in the revision they are checked out. If you move in history, it lazy-downloads missing local blobs automatically.

Thanks for the reference.  This doesn't directly answer my question about what would happen specifically when each step of Braid's workflow runs on a blobless clone; a bit more research will be needed.

> > (3) Understand any potential interactions with other enhancements we're considering that might be higher priority than this one.  The obvious one to me is [this proposal](https://github.com/cristibalan/braid/issues/71#issuecomment-697295152) to avoid copying all the vendor repository data into the downstream repository.  Another such enhancement might be to use a shallow clone in the cache repository.
> 
> Shallow clones have only partial history, so some merges can fail.

OK.  I'll want to better understand the trade-offs between shallow clones and blobless clones.  I don't think the issue you cited with merges is relevant to Braid: Braid only performs merges where the merge base is explicitly specified and not inferred from the history in a way that can fail with a shallow clone.  In the extreme case where the user adds a mirror at a given upstream revision and never updates it (and thus never needs any upstream content other than that single revision), I assume a shallow clone would perform better than a non-shallow blobless clone.  However, as users work with more upstream revisions, the trade-off would shift in favor of the blobless clone, if I understand correctly.

In summary, I'm not going to accept a PR like this without further work/research on each of my points (1)-(4).  This won't be a priority for me any time soon, so we'll probably need your help if you want to get this merged.

One alternative worth considering may be to add an opt-in flag for Braid to use a blobless clone and document that the feature is considered experimental.  Then the standard of vetting we'd need is much lower than for a change that is imposed on all users of Braid.

**Review by mattmccutchen — CHANGES_REQUESTED** [review]
Marking this as "changes requested" so it's obvious from the repository PR list that we've responded to the PR, even though what's needed may consist more of research than actual changes to the code of the PR.

**Comment by yajo** [comment]
Thanks for the reviews. In the mean time, I developed my own solution, [Mr. Chef](https://gitlab.com/moduon/mrchef). I hope you understand I'll just close this PR. Feel free to take the code and do whatever you want with it if you feel it's valuable.

Regards.
